### PR TITLE
Rough GLSL syntax

### PR DESCRIPTION
**Pull request by Codysseus**
Here is a basic syntax representation. Some things are redundant
when translating. Please let me know what you think.

### CODE REVIEW

File: src/GLSL.hs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
+           | GSwitch GExpr [(GConst,[GStmt])]
+           | GStmt GExpr
+           | GRet GExpr
+           | GBreak
+           | GContinue
+
+data GTop = GFunc String [(String,GType)] GType [GStmt]
+          | GGlob GExpr                     -- This is for global variables
```
Comment (montymxb) [inline src/GLSL.hs:72]: I believe in GLSL (i may have to check this) you can only declare globals as names with associated types with or without a corresponding binding. With the way this is written out we can write arbitrary expressions like binary ops at the top level w/out being part of a binding. What do you think if we created a new category for global vars kinda like this (also a rough idea).
```haskell
-- top level declarations, funcs, global vars, or opaque types.
data GTop = GGlob Global | ...

-- globals we can write out
data Global = GlobalDecl Name GType -- declarations w/ no binding
                     | GlobalDef Name GType GExpr  -- declarations w/ binding (a def)
```

File: src/GLSL.hs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
+
+data GStmt = GFor GExpr GExpr GExpr [GStmt]
+           | GWhile GExpr [GStmt]
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
```
Comment (montymxb) [inline src/GLSL.hs:64]: We should enforce that if & else-if must always have a condition, and must always have at least one branch (possibly followed by a branch w/ no condition)
```haskell
type CondBranch = (Cond,[Stmt])

-- a bit crude, but then we have the idea that a
-- branch is minimum 1 branch w/ statements guarded by a condition,
-- followed by zero or more condtional branches & maybe a final `else`
data GStmt = ...
     | GIf CondBranch [CondBranch] (Maybe [Stmt])
```

File: src/GLSL.hs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            | GV [GConst]
+            | GM [[GConst]]
+
+data GExpr = GCall String [GExpr]
+           | GParen GExpr
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
```
Comment (montymxb) [inline src/GLSL.hs:59]: I think working with arrays could qualify as binary ops, but the lefthand side always needs to be a name for that to work. A single case for indexing a name would work, with the following expr resolving to either a var or an int.

File: src/GLSL.hs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+data GStmt = GFor GExpr GExpr GExpr [GStmt]
+           | GWhile GExpr [GStmt]
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
+           | GSwitch GExpr [(GConst,[GStmt])]
+           | GStmt GExpr
+           | GRet GExpr
+           | GBreak
+           | GContinue
```
Comment (montymxb) [inline src/GLSL.hs:69]: Vertext & Fragment shaders can have different kinds of statements. The most noteworthy example is the `discard` statement, which discards a given fragment leaving it 'empty' in a sense. This is used to achieve certain effects, like putting holes in a surface or something similar. To incorporate this, we would at least 3 categories of statements.
```
Stmt -- general statements shared by all shaders
VStmt -- statements only available in the vertex shader (I have to check if there are any...)
FStmt -- statements only available in the fragment shader (at least discard, possibly others)
```

File: src/GLSL.hs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+data GTop = GFunc String [(String,GType)] GType [GStmt]
+          | GGlob GExpr                     -- This is for global variables
+          | GOpaque String OpaqueType GType -- This is for varying and Attribute
+
+type GShader = [GTop]
```
Comment (montymxb) [inline src/GLSL.hs:75]: Similar to what I mentioned above about different kinds of statements, we would probably break this into a data type with constructors for Vertex vs. Fragment shaders that limit that allow use of the appropriate statements in the right stages. For now, that's just discard, but I believe there are others as well (layout comes to mind).

File: src/GLSL.hs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        deriving (Eq, Ord)
+
+data GType = Void
+           | GBool
+           | GInt
+           | GUInt
+           | GFloat
+           | GDouble
+           | GVec GSize GType
+           | GMat GSize GSize
```
Comment (montymxb) [inline src/GLSL.hs:16]: Nice 👍 , reminds me of dependent types!

**Review by montymxb — COMMENTED** [review]
This is looking good for a GLSL syntax! I left quite a lot of notes, some questions, maybe some things that I missed too so feel free to correct me.

We should also add a section for declaring precision of certain GLSL data types, but this is important only for GLES. Which brings up whether or not we want to include pre-processor directives here. They're important, as they indicate which version of GLSL should be targeted, and alter what code is produced, but they're not really part of the language itself (just the concrete syntax). However, if we expect to be able to pretty-print working GLSL programs from this representation, we need to include those details somewhere, as it's likely we'll want to target different versions GLSL down the road.

File: src/GLSL.hs (hunk 1 above)
Comment (Codysseus) [reply src/GLSL.hs:72]: I think this begs the question as to how much of the full specification do we want to encompass?
I agree with you and was wondering that myself when I was writing it out, but my thought was we could enforce that it's only assignments at the type level (but that's probably poor form, I know).

File: src/GLSL.hs (hunk 2 above)
Comment (Codysseus) [reply src/GLSL.hs:64]: Aaaah very clever and  a more correct way to handle this. Though here you don't define `Cond`. Should this be a lower nonterminal than expression?

File: src/GLSL.hs (hunk 3 above)
Comment (Codysseus) [reply src/GLSL.hs:59]: Yes, if you look at the `GOp` list I had considered that! Though you make a good point that it must always be a name. I had given some thought to make the `GOp` list a GADT so as to enforce this at the Haskell type level, but then I remembered the issue with references.

Comment (Codysseus) [reply src/GLSL.hs:59]: I was playing around and found you could alter type annotations through a function like this
```haskell
trans :: Expr a -> b -> Expr b
trans (Ref e) _ = Ref e
```
Reminds me of writing Idris!

**Comment by Codysseus** [comment]
This PR will likely stall as we are considering using GLSL syntax defined in other projects.

**Comment by Codysseus** [comment]
We are using the Rust glsl library.

**Pull request closed without merging**